sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs

```diff
@@ -156,12 +170,18 @@ public override void Clear()
         ///
         public override IReadOnlyCollection<T> AsReadOnly<T>()
         {
-            if (typeof(T) != typeof(ServiceBusMessage))
+            if (typeof(T) == typeof(AmqpMessage))
+            {
+                return (IReadOnlyCollection<T>)BatchMessages;
+            }
+            else if (typeof(T) == typeof(ServiceBusMessage))
```

Comment:
Because of `DiagnosticScope` and needing the first `ServiceBusMessage` to populate the batch envelope, I wasn't sure how to make these changes work without keeping the `ServiceBusMessage` references in the `AmqpMessageBatch`. I am open to other suggestions in this regard. 

I considered changing `DiagnosticScope` (but at first investigation it seems like it would be a pretty extensive change) and `AmqpMessageBatch` (but since there are so many properties held in the `ServiceBusMessage`, a lot would need to be propagated from the first message, unlike Event Hubs which just needs the partition key).